### PR TITLE
Import from Trello

### PR DESCRIPTION
**Pull request by vitormattos**
* Resolves: #87 
* Target version: master 

### Summary

Command to import json from Trello boards to Deck.

> **PS**: The import command is ready to accept import from other systems and not just Trello. To create import from other systems it will be necessary to create a class for the system to be imported.

**Command help**:

```
deck:import --help
Description:
  Import data

Usage:
  deck:import [options]

Options:
      --system=SYSTEM   Source system for import. Available options: Trello API, Trello JSON.
      --config=CONFIG   Configuration json file. [default: "config.json"]
      --data[=DATA]     Data file to import. [default: "data.json"]
```

**API documentation**
See `docs/API.md`

### TODO

- Import flow
  - [x] Import board
  - [x] Import labels
  - [x] Import stacks
  - [x] Import cards
  - [x] Associate cards to labels
  - [x] Import comments
  - [x] Import attachments. Attachment URL added in the attachment table;
  - [x] Import user assigment
    - [x] Relationship of Trello users to Nextcloud users
- Import systems
  - [x] Import from Trello JSON (limited to 1000 Trello actions)
  - [x] Import from Trello API (no data limits to import)
  - [x] Implement architecture to make possible create an importer to other systems
- Implement endpoints to import using Deck API
  - [x] Endpoint to import using API
  - [x] Endpoint to get available systems to import
  - [x] Endpoint to get json schema of a system to import

### Checklist

- [x] Code is properly formatted
- [x] Sign-off message is added to all commits
- [x] Tests (unit, integration, api and/or acceptance) are included
- [x] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

**Comment by vitormattos** [comment]
Finally done!

Code coverage with tests is not ideal but I think it is acceptable.

The breaking tests are related to the setup of the server, it seems to me it has no relationship with this PR.

NOTE: It seems a good thing to remove support for PHP 7.2 and 7.3, I had to remove the beautiful arrow functions and heredoc.

**Comment by vitormattos** [comment]
Hi  @juliushaertl,

I organized the code and I think it is now more beautiful and ready to accept other import sources.

I didn't implement the frontend in VueJS for importing but it already has the API endpoints ready to be consumed and working.

I thought about implementing Asana import but I've committed too much in this PR :-D

**Comment by juliushaertl** [comment]
First of all thanks a lot for your contribution, this is really awesome and looks very promising from a first quick look. 

It might take a bit until I get to fully review it, but I'll dig into it in detail as soon as possible.

> NOTE: It seems a good thing to remove support for PHP 7.2 and 7.3, I had to remove the beautiful arrow functions and heredoc.

We basically stick with what is compatible in the Nextcloud release, which is currently requiring 7.3 as a minimum PHP version. We can indeed drop 7.2 for any newer branch than stable1.2.

**Comment by vitormattos** [comment]
@juliushaertl  to make code review easier, I updated the documentation with the class diagram.

I used the extension [yUML](https://github.com/jaime-olivares/vscode-yuml) for the VSCode|Codium.

**Comment by vitormattos** [comment]
Hi @juliushaertl,

I returned to draft to implement import using Trello API.
It's ready for code review again, I updated the documentation too.

**Review by eneiluj — CHANGES_REQUESTED** [review]
Great work!

Could you provide a test JSON file to import?

If I understand correctly, when importing from Trello API, all data is first loaded and stored in `BoardImportTrelloApiService` attributes. Once everything is loaded, Deck entities are added into the DB. I think there is a chance to reach the max RAM usage when importing big Trello projects.

I don't know how hard it would be to make those changes but: Would it be possible to import progressively while the data is loaded?
It would be very effective in saving RAM usage, even more thanks to the request pagination. Data could be added after each `doRequest` call.

**Comment by eneiluj** [comment]
Oh and also, you need to rebase this branch on master. You could also squash the commits into one (or up to 5 or 6, less than 10 anyway :grin:) to keep the history clean.

File: lib/Controller/BoardImportApiController.php
Comment (vitormattos) [reply lib/Controller/BoardImportApiController.php:59]: The config variable has associative array and objects.

`json_decode` without the second parameter will make everything object.

File: lib/Service/BoardImportTrelloApiService.php
Comment (vitormattos) [reply lib/Service/BoardImportTrelloApiService.php:156]: This method returns the body for API requests. In some cases it is a collection (array) and in others it is an object. It depends on which endpoint the request is for.

**Comment by vitormattos** [comment]
@eneiluj 
> Could you provide a test JSON file to import?

See on [tests/data/data-trelloJson.json](https://github.com/nextcloud/deck/pull/3182/files/06f272a64020f5d3fbc62ed66c2e684efaab65f4#diff-8abf3dfd178c96bb5a3a5f6c0f9d167979dd9e50417b488da08a38cf897fef06)

**Comment by vitormattos** [comment]
On my user case I imported a board with more than 1000 cards and comments with more than 1000 characters and it worked.

It's possible that a very big board could overflow memory but I don't know how big.

Trello works with collections, for the import to work need to be in memory to join them. It is possible to think of some approach to splitting these collections into smaller files but this will increase the code complexity. Splitting in smaller files it will be necessary to have files with the search indexes.

> If I understand correctly, when importing from Trello API, all data is first loaded and stored in `BoardImportTrelloApiService` attributes. Once everything is loaded, Deck entities are added into the DB. I think there is a chance to reach the max RAM usage when importing big Trello projects.
> 
> I don't know how hard it would be to make those changes but: Would it be possible to import progressively while the data is loaded?
> It would be very effective in saving RAM usage, even more thanks to the request pagination. Data could be added after each

**Comment by eneiluj** [comment]
Thanks for the answers.

> for the import to work need to be in memory to join them

Right, the whole data has to be loaded because there are relations between entities. Alright then, fine :grin:.

Could you be a little more explicit in BoardImportCommandService::validateConfig() about failure reasons? The error is the same whether the file does not exists or it couldn't be decoded or it's not a valid config.

I'm sorry, this won't be easy but you need to rebase your work on top of the master branch. While working on long PRs like this one, it's better to regularly rebase than to regularly merge so all your work is on top and the PR only adds stuff at the end of the commit history.
Once you have rebased, you can squash some of your commits together to make the history a little bit cleaner. You can do that while being on your branch with `git rebase -i master`.

**Comment by vitormattos** [comment]
@eneiluj  thank you very much about how to use rebase command

**Comment by eneiluj** [comment]
It seems fine to me. The amount of commit is now more reasonable :grin:.
As I don't have a deep knowledge of the Deck app. Let's wait for a review from @juliushaertl.
Once it's accepted, just rebase once more on master and you should be able to merge this.

**Comment by vitormattos** [comment]
ping @juliushaertl

File: lib/Service/BoardImportService.php
Comment (eneiluj) [reply lib/Service/BoardImportService.php]: I agree with @juliushaertl, it would be better to import the classes statically at the top of this file and a static list in this method.
I would say it's ok to add the event emission and handling in a future PR.

Comment (vitormattos) [reply lib/Service/BoardImportService.php]: I will dispatch event on this place. What you think @eneiluj and @juliushaertl ?

**Review by eneiluj — APPROVED** [review]
Fine by me! :+1: Amazing work

**Comment by vitormattos** [comment]
The broken tests on PHP 7.3 was fixed by #3498

**Comment by juliushaertl** [comment]
Rebased and resovled the composer.lock conflict

**Review by juliushaertl — APPROVED** [review]

**Comment by juliushaertl** [comment]
Great work, thanks a lot @vitormattos :clap:

**Pull request merged**
